feat(go): Schema definition system with dotprompt

ihan211 opened the pull request:
This PR adds a schema definition system to genkit with Dotprompt reference support.
Key Components

    Schema registration with Dotprompt framework
    DefineSchema API for defining prompt data structures
    Enhanced initialization process for schema registration
    Schema resolution within templates

Example Usage

// Define a product schema
type ProductSchema struct {
    Name        string  `json:"name"`
    Description string  `json:"description"`
    Price       float64 `json:"price"`
    Category    string  `json:"category"`
    InStock     bool    `json:"inStock"`
}

// Register the schema with genkit
genkit.DefineSchema("ProductSchema", ProductSchema{})

// Reference the schema in a Dotprompt template
// promptContent example:
// ---
// input:
//   schema:
//     theme: string
// output:
//   schema: ProductSchema
// ---
// Generate a product that fits the {{theme}} theme.
// Make sure to provide a detailed description and appropriate pricing.

The implementation enables:
* Automatic schema registration during Genkit initialization
* Structured output parsing from model responses to Go structs
* Schema reference resolution within Dotprompt templates
* JSON Schema support for standardized type definitions

Testing
Unit tests in schema_test.go
Working example in samples directory

Checklist (if applicable):
- [X] PR title is following https://www.conventionalcommits.org/en/v1.0.0/
- [X] Tested (manually, unit tested, etc.)
- [ ] Docs updated (updated docs or a docs bug requir

ihan211 committed:
feat(go/genkit): add DefineSchema and Dotprompt reference support

ihan211 committed:
feat(go/genkit): Refactor schema handling to resolve import cycles and improve code organization

- Move pendingSchemas management to core package to avoid validation duplication
- Add RegisterGlobalSchemaResolver that follows DRY by calling registerSchemaResolver
- Create recursive helper function for field-to-schema conversion to support nested structs
- Fix import cycles between genkit, core, and registry packages
- Standardize schema path construction with consistent separators
- Add functions to retrieve and clear pending schemas
- Update schema registration during initialization process
- Fix LookupPrompt function signature to match usage

This refactoring addresses code review feedback by:
1. Eliminating duplicate validation between genkit.DefineSchema and core.RegisterSchema
2. Following DRY principle for schema resolver registration
3. Improving support for nested struct types with recursive field conversion
4. Creating a cleaner architecture with proper separation of concerns

ihan211 committed:
feat(go/genkit): Fix dependency conflict with dotprompt by adding replace directive

ihan211 committed:
Update dotprompt dependency to use add-schema-support branch

ihan211 committed:
feat(go/genkit): Refactor schema registry: rename Get prefixed functions and improve naming convention

hugoaguirre commented on go/genkit/schema.go:
```diff
@@ -0,0 +1,117 @@
+	// Register with core registry
+	core.RegisterSchema(name, schema)
+
+	// Also track for Dotprompt integration
```
I think you can change this comment to explain `why` is needed to keep track of the pending schema in dotprompt instead of mentioning `what` is being done

hugoaguirre commented on go/genkit/schema.go:
```diff
@@ -0,0 +1,117 @@
+		panic("genkit.DefineSchema: schema cannot be nil")
+	}
+
+	// Register with core registry
```
nit: not needed

hugoaguirre commented on go/genkit/schema.go:
```diff
@@ -0,0 +1,117 @@
+		return jsonSchema
+	}
+
+	// Register the resolver with Dotprompt
```
we can ignore this comment, same for the one at line 112

hugoaguirre commented on go/genkit/schema.go:
```diff
@@ -0,0 +1,117 @@
+	schemaResolver := func(name string) any {
+		schema, exists := LookupSchema(name)
+		if !exists {
+			fmt.Printf("Schema '%s' not found in registry\n", name)
```
Use `slog` to log these kind of messages, here's an example of how it is used:
```go
slog.Error("schema not found in registry", "name", name)
```

ihan211 replied:
replaced

hugoaguirre commented on go/internal/registry/schema.go:
```diff
@@ -0,0 +1,236 @@
+
+	r.Dotprompt.DefineSchema(name, jsonSchema)
+	r.RegisterValue(SchemaType+"/"+name, structType)
+	fmt.Printf("Registered schema '%s' with registry and Dotprompt\n", name)
```
If these logs are not for debugging purposes, use `slog` instead, otherwise I think you should remove them

ihan211 replied:
removed.

hugoaguirre commented on go/internal/registry/schema.go:
```diff
@@ -0,0 +1,236 @@
+}
+
+// DumpRegistrySchemas prints all schemas stored in the registry
+func (r *Registry) DumpRegistrySchemas() {
```
I cannot find where this function is used, why is it needed?

ihan211 replied:
removed.

hugoaguirre commented on go/internal/registry/schema.go:
```diff
@@ -0,0 +1,236 @@
+
+// convertStructToJsonSchema converts a Go struct to a JSON schema
+func convertStructToJsonSchema(structType any) (*jsonschema.Schema, error) {
+	fmt.Printf("Converting schema of type %T to JSON Schema\n", structType)
```
nit: not needed, we can remove this

ihan211 replied:
removed.

ihan211 committed:
fix(go/genkit): address issues raised by @hugoaguirre

- Move ClearSchemas to schema_test.go as clearSchemasForTest()
- Replace panic with error returns in RegisterSchema and DefineSchema
- Improve logging using structured slog instead of fmt.Printf
- Remove debug prints and unused functions
- Add proper error handling for schema not found cases

ihan211 commented on go/core/schema.go:
```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: Apache-2.0
+// Package core provides core functionality for the genkit framework.
+
+package core
```
done

ihan211 commented on go/core/schema.go:
```diff
@@ -0,0 +1,142 @@
+
+// ClearSchemas removes all registered schemas.
+// This is primarily for testing purposes.
+func ClearSchemas() {
```
moved to test file.

ihan211 commented on go/core/schema.go:
```diff
@@ -0,0 +1,142 @@
+// Returns the schema for convenience in chaining operations.
+func RegisterSchema(name string, schema any) Schema {
+	if name == "" {
+		panic("core.RegisterSchema: schema name cannot be empty")
```
replaced.

ihan211 commented on go/genkit/genkit.go:
```diff
@@ -245,6 +250,22 @@ func Init(ctx context.Context, opts ...GenkitOption) (*Genkit, error) {
 
+// Internal function called during Init to register pending schemas
+func registerPendingSchemas(reg *registry.Registry) error {
+	// Get pending schemas from core
```
removed.

ihan211 commented on go/genkit/schema.go:
```diff
@@ -0,0 +1,117 @@
+//	personSchema := genkit.DefineSchema("Person", Person{})
+func DefineSchema(name string, schema Schema) Schema {
+	if name == "" {
+		panic("genkit.DefineSchema: schema name cannot be empty")
```
replaced.
